client/history/client.go

```diff
@@ -638,6 +645,57 @@ func (c *clientImpl) SyncActivity(
 	return err
 }
 
+func (c *clientImpl) GetReplicationTasks(
+	ctx context.Context,
+	request *replicator.GetReplicationTasksRequest,
+	opts ...yarpc.CallOption,
+) (*replicator.GetReplicationTasksResponse, error) {
+	requestsByClient := make(map[historyserviceclient.Interface]*replicator.GetReplicationTasksRequest)
+
+	for _, token := range request.Tokens {
+		client, err := c.getClientForShardID(int(token.GetShardID()))
+		if err != nil {
+			return nil, err
+		}
+
+		if _, ok := requestsByClient[client]; !ok {
+			requestsByClient[client] = &replicator.GetReplicationTasksRequest{}
+		}
+
+		req := requestsByClient[client]
+		req.Tokens = append(req.Tokens, token)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(len(requestsByClient))
+	respChan := make(chan *replicator.GetReplicationTasksResponse, len(requestsByClient))
+	for client, req := range requestsByClient {
+		go func(client historyserviceclient.Interface, request *replicator.GetReplicationTasksRequest) {
+			ctx, cancel := c.createContext(ctx)
```

Comment:
Do we have to create new context for every one?

Reply:
yes, otherwise one cancel will cancel all the requests, right?

Reply:
You can put defer cancel() at the end of GetReplicationTasksI(). It is for preventing memory leak. But all the requests can use the same context because they are using the same timeouts.